sdks/python/apache_beam/runners/direct/transform_evaluator.py

```diff
@@ -634,6 +635,7 @@ def process_element(self, element):
 
   def finish_bundle(self):
     self.runner.finish()
+    self.runner.teardown()
```

Comment:
I think `__del__` is a sort of right idea for teardown. You could do better but you _must not_ call any method of it after you call teardown. So that is why it is `__del__`.

Reply:
I would prefer to be explicit here with `teardown` because `__del__` could be called in many cases, and it could be (for example for testing) that these instances are not being explicitly used during execution when it is being garbage collected.

Reply:
+1 for making this an explicit call.